test/scale/scale_tester.go

```diff
@@ -0,0 +1,358 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scale
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	scale "k8s.io/api/autoscaling/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"sigs.k8s.io/azuredisk-csi-driver/test/e2e/driver"
+	testtypes "sigs.k8s.io/azuredisk-csi-driver/test/types"
+	nodeutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/node"
+	podutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/pod"
+	volutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/volume"
+)
+
+type PodSchedulingWithPVScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+type PodSchedulingOnFailoverScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	PodCount               int
+	StorageClassParameters map[string]string
+}
+
+type PodSchedulingOnFailoverMultiplePVScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	PodCount               int
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+func (t *PodSchedulingWithPVScaleTest) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	totalNumberOfPods := t.Replicas
+	var oneMinPodsRunning,
+		twoMinPodsRunning,
+		threeMinPodsRunning,
+		fiveMinPodsRunning,
+		tenMinPodsRunning = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+	var oneMinPodsDeleted,
+		twoMinPodsDeleted,
+		threeMinPodsDeleted,
+		fiveMinPodsDeleted,
+		tenMinPodsDeleted = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+	startScaleOut := time.Now()
+	// Get the list of available nodes for scheduling the pod
+	nodes := nodeutil.ListNodeNames(client)
+	if len(nodes) < 3 {
+		ginkgo.Skip(fmt.Sprintf("need at least 3 nodes to verify the test case. Current node count is %d", len(nodes)))
+	}
+	tStorageClass, storageCleanup := t.Pod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+	statefulSet, cleanupStatefulSet := t.Pod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, t.Replicas, t.StorageClassParameters, &tStorageClass)
+	for i := range cleanupStatefulSet {
+		i := i
+		defer cleanupStatefulSet[i]()
+	}
+
+	statefulSet.CreateWithoutWaiting()
+	numberOfRunningPods := 0
+	for {
+		numberOfRunningPods = podutil.CountAllPodsWithMatchingLabel(client, namespace, map[string]string{"app": "azuredisk-volume-tester"}, string(v1.PodRunning))
+		if numberOfRunningPods >= totalNumberOfPods || time.Since(startScaleOut) > 30*time.Minute {
```

Comment:
It would be good to specify the test timeout through a command line option rather than hardcoding it.